packages/tpetra/core/src/Tpetra_withLocalAccess_MultiVector.hpp

```diff
@@ -56,309 +54,462 @@
-            LA.G_.clear_sync_state ();
+        if (! LA.isValid()) { // return "null" Kokkos::View
+          return master_local_view_type();
+        }
+        else {
+          using access_mode = typename local_access_type::access_mode;
+          constexpr bool is_write_only =
+            std::is_same<access_mode, write_only>::value;
+          if (is_write_only) {
+            LA.G_.clear_sync_state();
           }
 
-          // mfh 06 Jun 2019: It could be that
-          // Kokkos::DefaultHostExecutionSpace != the Vector's
-          // execution_space.  For example, the first could be
-          // Kokkos::OpenMP, but the second could be Kokkos::Serial.
-          // That's why we go through the trouble below.
+          // Given that Tpetra::(Multi)Vector currently uses
+          // Kokkos::DualView, here is how get() must behave:
           //
-          // It's easier to use an execution space than a memory
-          // space.  Otherwise, DualView of CudaUVMSpace complains
-          // that HostSpace is not one of its two memory spaces.
-          // (Both the device and the host Views of a DualView of
-          // CudaUVMSpace have memory_space = CudaUVMSpace.)
-          using space = typename std::conditional<is_host,
-            typename dual_view_type::t_host::execution_space,
-            typename dual_view_type::t_dev::execution_space>::type;
-
-          if (LA.G_.template need_sync<space> ()) {
-            LA.G_.template sync<space> ();
+          // - LA's memory space tells us which allocation to view.
+          //
+          // - LA's execution space tells us whether we need to fence.
+          //
+          //   - If LA's execution space equals the MultiVector's
+          //     execution space, then there is no need to fence, no
+          //     matter what the requested memory space is.
+          //
+          //   - Else, if LA's execution space is a host space, but
+          //     the MultiVector needs sync to host, then we must
+          //     fence the MultiVector's (device) execution space, to
+          //     ensure that device kernels aren't concurrently
+          //     modifying the MultiVector's local data.
+          //     Tpetra::MultiVector::fence (should) do that for us,
+          //     via Kokkos::DualView::fence.
+
+          // It's easier to use an execution space than a memory space
+          // in sync<Space>.  Otherwise, DualView of CudaUVMSpace
+          // complains that HostSpace is not one of its two memory
+          // spaces.  (Both the device and the host Views of a
+          // DualView of CudaUVMSpace have memory_space =
+          // CudaUVMSpace.)  Furthermore, this handles the case where
+          // Kokkos::DefaultHostExecutionSpace != the MultiVector's
+          // execution space, but the latter is still a host space
+          // (e.g., Kokkos::OpenMP vs. Kokkos::Serial).
+
+          if (LA.G_.template need_sync<execution_space>()) {
```

Comment:
The key here is that we now use the user's specified `execution_space`, if the user specifies one.  The key use case is that the Tpetra::MultiVector uses CudaUVMSpace to store data, but the user wants to access at `DefaultHostExecutionSpace`.